webrender/src/device/gfx/device.rs

```diff
@@ -1007,40 +1016,73 @@ impl<B: hal::Backend> Device<B> {
 
         let frame_count = swap_config.image_count as usize;
 
-        let (swap_chain, images) =
-            unsafe { device.create_swapchain(surface, swap_config, old_swap_chain) }
-                .expect("create_swapchain failed");
+        println!("## Image count {}", frame_count);
 
-        assert_eq!(images.len(), frame_count);
+        unsafe {
+            surface.configure_swapchain(&device, swap_config)
+                .expect("Can't configure swapchain");
+        };
 
-        let image_extent = hal::image::Extent {
-            width: window_extent.width as _,
-            height: window_extent.height as _,
-            depth: 1,
+        let viewport = hal::pso::Viewport {
+            rect: hal::pso::Rect {
+                x: 0,
+                y: 0,
+                w: window_extent.width as _,
+                h: window_extent.height as _,
+            },
+            depth: 0.0 .. 1.0,
         };
 
-        let (frames, viewport) =
-            Self::init_frames(
-                device,
-                heaps,
-                image_extent,
-                render_passes,
-                available_surface_format,
-                frame_count,
-                Some(images),
-            );
+        let surface_image = unsafe {
+            match surface.acquire_image(!0) {
+                Ok((image, _)) => image,
+                Err(_) => unimplemented!(),
+            }
+        };
+
+        let depth = DepthBuffer::new(
+            device,
+            heaps,
+            window_extent.width,
+            window_extent.height,
+            DEPTH_FORMAT,
+        );
+
+        let frame = Frame::new(
+            device,
+            surface_image,
+            &depth.core.view,
+            viewport.rect,
+            render_passes,
+        );
+
+        let mut frame_depths = ArrayVec::new();
+        frame_depths.push(depth);
+        for _ in 1 .. frame_count {
+            frame_depths.push(
+                DepthBuffer::new(
```

Comment:
why are we creating N depth buffers? we should be able to use the same depth buffer for all frames
moreover, with picture caching we may not even need any depth buffers for the swapchain images

Reply:
As I've found picture caching uses the depth buffer.